src/renderer/stores/chat.ts

```diff
@@ -843,12 +849,76 @@ class ChatStore extends Store<ChatStoreState> {
             throw new Error('no account set')
           }
           // this function already throws an error if message is not found
-          const message = await BackendRemote.rpc.messageGetMessage(
-            accountId,
-            msgId
-          )
 
-          const chatId = (message as Type.Message).chatId
+          let chatId = -1
+          let jumpToMessageId = -1
+          let jumpToMessageStack: number[] = []
+          let message: Type.Message | undefined = undefined
+          if (msgId === undefined) {
+            const jumpToMessageStackLength = this.state.jumpToMessageStack
+              .length
+            if (jumpToMessageStackLength !== 0) {
+              jumpToMessageStack = this.state.jumpToMessageStack.slice(
+                0,
+                jumpToMessageStackLength - 1
+              )
+              jumpToMessageId = this.state.jumpToMessageStack[
+                jumpToMessageStackLength - 1
+              ]
+              message = await BackendRemote.rpc.messageGetMessage(
+                accountId,
+                jumpToMessageId as number,
+              )
+              chatId = message.chatId
+
+            } else {
+              jumpToMessageStack = []
+              jumpToMessageId = this.state.messageIds[
+                this.state.messageIds.length - 1
+              ]
+              message = await BackendRemote.rpc.messageGetMessage(
+                accountId,
+                jumpToMessageId
+              )
+              chatId = message.chatId
+              return this.effect.selectChat(chatId)
+            }
+          } else if (addMessageIdToStack === undefined) {
+            // reset jumpToMessageStack
+            message = await BackendRemote.rpc.messageGetMessage(
+              accountId,
+              msgId as number,
+            )
+            chatId = message.chatId
+
+            jumpToMessageId = msgId as number
+            jumpToMessageStack = []
+          } else {
+            message = await BackendRemote.rpc.messageGetMessage(
+              accountId,
+              msgId as number,
+            )
+            chatId = message.chatId
+
+            jumpToMessageId = msgId as number
+            // If we are not switching chats, add current jumpToMessageId to the stack
+            const currentChatId = chatStore.state.chat?.id || -1
+            if (chatId !== currentChatId) {
+              jumpToMessageStack = []
+            } else if (chatStore.state.jumpToMessageStack.indexOf(addMessageIdToStack) !== -1) {
+              jumpToMessageStack = chatStore.state.jumpToMessageStack
+            } else {
+              jumpToMessageStack = [...chatStore.state.jumpToMessageStack, addMessageIdToStack]
+            }
+          }
```

Comment:
I'm not really happy with this part, but also don't know a way of improving that code.